### PR TITLE
Unix profile selector

### LINKED ISSUE #118999: Implement default profiles for linux/osx
Follow up from #118790:

Deferred creating the default profiles for linux/osx to get the PR merged/tested asap.

### LINKED ISSUE #119179: Refine default quick launch profile list on non-Windows
We want to only show bash, zsh, fish, tmux if available:

Screen Shot 2021-03-17 at 5 16 59 AM

### PR DESCRIPTION
**Pull request by meganrogge**
This PR 
fixes #119179
fixes #118999

tested using custom build on WSL
recording (54)

### CODE REVIEW

File: src/vs/workbench/contrib/terminal/node/terminalProfiles.ts [hunk 1]
```diff
@@ -198,15 +198,31 @@ async function getWslProfiles(wslPath: string, detectWslProfiles?: boolean, logS
...
 			profileName: basename(e),
 			path: e
 		};
 	});
+	if (!quickLaunchOnly) {
+		return detectedProfiles;
+	}
+	let validProfiles: ITerminalProfile[] = [];
+
+	for (const [profileKey, value] of Object.entries(configProfiles)) {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/node/terminalProfiles.ts:215]: I think the use of `Object.entries` here is why `Map` is working:

Screen Shot 2021-03-17 at 4 11 16 PM

The type isn't a `Map` as this wouldn't work:

Screen Shot 2021-03-17 at 4 11 58 PM

File: src/vs/workbench/contrib/terminal/node/terminalProfiles.ts [hunk 2]
```diff
@@ -198,15 +198,31 @@ async function getWslProfiles(wslPath: string, detectWslProfiles?: boolean, logS
 	return [];
 }
 
-async function detectAvailableUnixProfiles(): Promise<ITerminalProfile[]> {
+async function detectAvailableUnixProfiles(quickLaunchOnly?: boolean, configProfiles?: any): Promise<ITerminalProfile[]> {
 	const contents = await fs.promises.readFile('/etc/shells', 'utf8');
 	const profiles = contents.split('\n').filter(e => e.trim().indexOf('#') !== 0 && e.trim().length > 0);
-	return profiles.map(e => {
+	const detectedProfiles = await profiles.map(e => {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/node/terminalProfiles.ts:204]: I don't think `await` is needed here

File: src/vs/workbench/contrib/terminal/node/terminalProfiles.ts [hunk 3]
```diff
@@ -198,15 +198,31 @@ async function getWslProfiles(wslPath: string, detectWslProfiles?: boolean, logS
...
 		return {
 			profileName: basename(e),
 			path: e
 		};
 	});
+	if (!quickLaunchOnly) {
+		return detectedProfiles;
+	}
+	let validProfiles: ITerminalProfile[] = [];
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/node/terminalProfiles.ts:213]:
```suggestion
	const validProfiles: ITerminalProfile[] = [];
```

File: src/vs/workbench/contrib/terminal/node/terminalProfiles.ts [hunk 4]
```diff
@@ -198,15 +198,31 @@ async function getWslProfiles(wslPath: string, detectWslProfiles?: boolean, logS
...
+
+	for (const [profileKey, value] of Object.entries(configProfiles)) {
+		if ((value as ITerminalExecutable).path) {
+			const configProfile = (value as ITerminalExecutable);
+			const path = configProfile.path;
+			const profile = Array.isArray(path) ? path.forEach(possiblePath => detectedProfiles.find(p => p.path.endsWith(possiblePath))) : detectedProfiles.find(p => p.path.endsWith(path));
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/node/terminalProfiles.ts:219]: I don't think `profile = path.forEach(...)` would work when it is an array?

File: src/vs/workbench/contrib/terminal/node/terminalProfiles.ts [hunk 5]
```diff
@@ -198,15 +198,31 @@ async function getWslProfiles(wslPath: string, detectWslProfiles?: boolean, logS
...
+			if (profile) {
+				validProfiles.push({ profileName: profileKey, path: profile.path });
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/node/terminalProfiles.ts:221]: Calling it profileName in the for would let us do use the shorthand:

```ts
validProfiles.push({ profileName, path: profile.path });
```

File: src/vs/workbench/contrib/terminal/test/node/terminalProfiles.test.ts [hunk 6]
```diff
@@ -9,7 +9,13 @@ import { detectAvailableProfiles, IStatProvider } from 'vs/workbench/contrib/ter
...
 	profiles: {
-		windows: {
+		windows?: {
+			[key: string]: { path?: string[], source?: string }
+		},
+		linux?: {
+			[key: string]: { path?: string[], source?: string }
+		},
+		osx?: {
 			[key: string]: { path?: string[], source?: string }
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/test/node/terminalProfiles.test.ts:19]: Should this pull in the profile interface used elsewhere?

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/node/terminalProfiles.ts (hunk 1 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/node/terminalProfiles.ts:215]: ok changed from `Map<string, ITerminalProfileObject>`  to ` { [key: string]: ITerminalProfileObject };`

Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/node/terminalProfiles.ts:215]: https://github.com/microsoft/vscode/blob/ff1ce589e7336c880152384d5f784dac0410f48b/src/vs/workbench/contrib/terminal/common/terminal.ts#L99-L99

File: src/vs/workbench/contrib/terminal/test/node/terminalProfiles.test.ts (hunk 6 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/test/node/terminalProfiles.test.ts:19]: yep, good point. now the test is failing though and i'm not sure why. wil linvestigate later per that issue

File: src/vs/workbench/contrib/terminal/node/terminalProfiles.ts [hunk 7]
```diff
@@ -198,15 +198,41 @@ async function getWslProfiles(wslPath: string, detectWslProfiles?: boolean, logS
...
 	});
+	if (!quickLaunchOnly) {
+		return detectedProfiles;
+	}
+	const validProfiles: ITerminalProfile[] = [];
+
+	for (const [profileName, value] of Object.entries(configProfiles)) {
+		if ((value as ITerminalExecutable).path) {
+			const configProfile = (value as ITerminalExecutable);
+			const path = configProfile.path;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/node/terminalProfiles.ts:218]: Maybe call this `pathOrPaths` so it's more obvious what the underlying type is?

File: src/vs/workbench/contrib/terminal/test/node/terminalProfiles.test.ts [hunk 8]
```diff
@@ -35,33 +32,35 @@ suite('Workbench - TerminalProfiles', () => {
...
-									source: 'Git Bash'
-								},
-							},
-						},
-						detectWslProfiles: false
-					};
-					const profiles = await detectAvailableProfiles(true, undefined, config, undefined, undefined, createStatProvider(_paths));
-					const expected = [{ profileName: 'Git Bash', path: _paths[0], args: ['--login'] }];
-					assert.deepStrictEqual(profiles, expected);
+					// const _paths = [`C:\\Program Files\\Git\\bin\\bash.exe`];
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/test/node/terminalProfiles.test.ts:35]: Is this being fixed up later? Need an issue?

**Review by Tyriar — APPROVED** [review]

File: src/vs/workbench/contrib/terminal/test/node/terminalProfiles.test.ts (hunk 8 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/terminal/test/node/terminalProfiles.test.ts:35]: there's an issue already

**Pull request merged**